src/calendar/Calendar.ts

```diff
@@ -0,0 +1,367 @@
+			if (key === (this._focusedDay + '-' + month)) {
+				element.focus();
+				this._callDateFocus = false;
+			}
+		}
+	}
+
+	@onPropertiesChanged()
+	protected onPropertiesChanged(evt: PropertiesChangeEvent<this, CalendarProperties>) {
+		const { customDateCell = CalendarCell } = this.properties;
+
+		// update custom option registry
+		if ( !this.registry || includes(evt.changedPropertyKeys, 'customDateCell')) {
+			this.registry = this._createRegistry(customDateCell);
+		}
+	}
+
+	private _createRegistry(customDateCell: any) {
+		const registry = new WidgetRegistry();
+		registry.define('date-cell', customDateCell);
+
+		return registry;
+	}
+
+	private _getMonthLength(month: number, year: number) {
+		const d = new Date(year, month + 1, 0);
+		return d.getDate();
+	}
+
+	private _getMonthYear() {
+		const {
+			month,
+			selectedDate = this._defaultDate,
+			year
+		} = this.properties;
+		return { month: typeof month === 'number' ? month : selectedDate.getMonth(), year: year ? year : selectedDate.getFullYear() };
+	}
+
+	private _goToDate(day: number) {
+		const {
+			month,
+			year
+		} = this._getMonthYear();
+		const currentMonthLength = this._getMonthLength(month, year);
+		const previousMonthLength = this._getMonthLength(month - 1, year);
+
+		if (day < 1) {
+			this._onMonthIncrement('down');
+			day += previousMonthLength;
+		}
+		else if (day > currentMonthLength) {
+			this._onMonthIncrement('up');
+			day -= currentMonthLength;
+		}
+
+		this._focusedDay = day;
+		this._callDateFocus = true;
+		this.invalidate();
+	}
+
+	private _onDateClick(date: number, disabled: boolean) {
+		const { onDateSelect } = this.properties;
+		let {
+			month,
+			year
+		} = this._getMonthYear();
+
+		if (disabled && date < 15) {
+			({ month, year } = this._onMonthIncrement('up'));
+			this._callDateFocus = true;
+		}
+		else if (disabled && date >= 15) {
+			({ month, year } = this._onMonthIncrement('down'));
+			this._callDateFocus = true;
+		}
+		this._focusedDay = date;
+		onDateSelect && onDateSelect(new Date(year, month, date));
+	}
+
+	private _onDateFocusCalled() {
+		this._callDateFocus = false;
+	}
+
+	private _onDateKeyDown(event: KeyboardEvent) {
+		const { month, year } = this._getMonthYear();
+		switch (event.which) {
+			case Keys.Up:
+				this._goToDate(this._focusedDay - 7);
+				break;
+			case Keys.Down:
+				this._goToDate(this._focusedDay + 7);
+				break;
+			case Keys.Left:
+				this._goToDate(this._focusedDay - 1);
+				break;
+			case Keys.Right:
+				this._goToDate(this._focusedDay + 1);
+				break;
+			case Keys.PageUp:
+				this._goToDate(1);
+				break;
+			case Keys.PageDown:
+				const monthLengh = this._getMonthLength(month, year);
+				this._goToDate(monthLengh);
+				break;
+			case Keys.Enter:
+			case Keys.Space:
+				const { onDateSelect } = this.properties;
+				onDateSelect && onDateSelect(new Date(year, month, this._focusedDay));
+		}
+	}
+
+	private _onMonthIncrement(direction: 'up' | 'down') {
```

Comment:
I'm good with that. Originally there was more repetitive logic involved so it made sense to only have one function. Now I think separating it might be best, and I like getting rid of `direction`